### PR TITLE
:bug: Print Info in Empty Repo Scans

### LINKED ISSUE #2157: Improving Score Reporting: Empty repository
Running Scorecard on an empty repository (no commits) returns results which do not make sense:

```
./scorecard --repo=azeemshaikh38/empty-repo-test --format=json | jq .
2022/08/17 01:00:47 unable to get tarball tarball not found: https://api.github.com/repos/azeemshaikh38/empty-repo-test/tarball/. Skipping...                                                        
{                                                                                                                                                                                                    
  "date": "2022-08-17",                                                                                                                                                                              
  "repo": {                                                                                                                                                                                          
    "name": "github.com/azeemshaikh38/empty-repo-test",                                                                                                                                              
    "commit": "no commits found"                                                                                                                                                                     
  },                                                                                                                                                                                                 
  "scorecard": {                                                                                                                                                                                     
    "version": "v4.5.0-25-g887facf",                                                                                                                                                                 
    "commit": "887facf3cad2bd401ff5f6a51725225f8ec28c17"                                                                                                                                             
  },                                                                                                                                                                                                 
  "score": 5,                                                                                                                                                                                        
  "checks": [                                                                                                                                                                                        
     ...
   ]
}
```

The aggregate score of 5 for an empty repository does not make any sense and needs to be fixed.

@ossf/scorecard-maintainers @spencerschrock @raghavkaul fyi.

### PR DESCRIPTION
**Pull request by leec94**
#### What kind of change does this PR introduce?

For empty repos, still return repository information, scorecard version, and time. This feature resolves #2157

- [x] PR title follows the guidelines defined in our [pull request documentation](https://github.com/ossf/scorecard/blob/main/CONTRIBUTING.md#pr-process)

#### What is the current behavior?
For empty repos, empty json is returned. 

#### What is the new behavior (if this is a feature change)?**

For empty repos, still return repository information, scorecard version, time, and score of -1. Local scans have not changed. 

- [x] Tests for the changes have been added (for bug fixes/features)

#### Which issue(s) this PR fixes

Fixes #2157

#### Special notes for your reviewer

Not sure if pkg/scorecard_test.go needs to be updated. How do I run this test to check? 

#### Does this PR introduce a user-facing change?

For user-facing changes, please add a concise, human-readable release note to
the `release-note`

(In particular, describe what changes users might need to make in their
application as a result of this pull request.)

```release-note
prints repository, scorecard version, and time data for scans of empty repos. no user action change required. 
```

### CODE REVIEW

File: pkg/scorecard.go [hunk 1]
```diff
@@ -105,9 +105,25 @@ func RunScorecard(ctx context.Context,
...
+				Name:      repo.URI(),
+				CommitSHA: commitSHA,
+			},
+			Scorecard: ScorecardInfo{
+				Version:   versionInfo.GitVersion,
+				CommitSHA: versionInfo.GitCommit,
+			},
+			Date: time.Now(),
+		}
+		return ret, nil
```
Comment (spencerschrock) [inline pkg/scorecard.go:122]: If we're not surfacing the error, let's not add it to the public error list. A private error at the top of this file should work too?

I was curious, so if we return the error we get something like this instead:
```
Error: RunScorecard: repository empty
2023/08/25 10:42:40 error during command execution: RunScorecard: repository empty
exit status 1
```

File: pkg/scorecard.go [hunk 2]
```diff
@@ -105,9 +105,25 @@ func RunScorecard(ctx context.Context,
...
+		ret := ScorecardResult{
+			Repo: RepoInfo{
+				Name:      repo.URI(),
+				CommitSHA: commitSHA,
+			},
+			Scorecard: ScorecardInfo{
+				Version:   versionInfo.GitVersion,
+				CommitSHA: versionInfo.GitCommit,
+			},
+			Date: time.Now(),
```
Comment (spencerschrock) [inline pkg/scorecard.go:120]: we have some code below that does the same thing. What do you think about moving it higher so we only define it in one spot?

File: pkg/scorecard.go [hunk 3]
```diff
@@ -105,9 +105,25 @@ func RunScorecard(ctx context.Context,
...
+		}
+		return ret, nil
+	} else if err != nil || commitSHA == "" {
```
Comment (spencerschrock) [inline pkg/scorecard.go:123]: do we still need the commitSHA comparison?

**Review by spencerschrock — COMMENTED** [review]
Some of the unit tests need fixed. You can check locally with `make unit-test`. Same with the linter: `make check-linter`

File: pkg/scorecard.go (hunk 1 above)
Comment (leec94) [reply pkg/scorecard.go:126]: changed this to a local error

File: pkg/scorecard.go (hunk 2 above)
Comment (leec94) [reply pkg/scorecard.go]: good idea. moved this higher too

File: pkg/scorecard.go (hunk 3 above)
Comment (leec94) [reply pkg/scorecard.go]: i removed this. hopefully its not needed in some random other edge case

**Comment by leec94** [comment]
@spencerschrock thanks for the feedback so far. i ran the tests and looks like there's still 1 where it's not passing, but i'm not sure how to get the `want` to match up with what's being returned. any advice?

File: pkg/scorecard.go (hunk 1 above)
Comment (spencerschrock) [reply pkg/scorecard.go:126]: can you make it private? e.g `errEmptyRepository`

**Comment by spencerschrock** [comment]
Repasting some of our Slack conversation here for others / so it doesn't go away:


So the test currently does:
```go
			if !reflect.DeepEqual(got, tt.want) {
				t.Errorf("RunScorecard() got = %v, want %v", got, tt.want)
			}
```
`reflect.DeepEqual` will compare every field. The timing issue is a common one in testing, and is a common example when dependency injection is talked about.

We use the `go-cmp` library elsewhere in our test code, which lets you do a better comparison:
https://pkg.go.dev/github.com/google/go-cmp/cmp#Equal
Here's an example from our code where we explicitly ignore a field. You will probably want to do that with the date field:
https://pkg.go.dev/github.com/google/go-cmp/cmp/cmpopts#IgnoreFields
https://github.com/ossf/scorecard/blob/63789835132192f60929f7a4511815536e7ec7b4/checks/evaluation/contributors_test.go#L149-L152

**Comment by leec94** [comment]
updated the test to ignore the date, since the seconds will always be off, and to check for "unknown" commitSHA. should pass unit-tests now! i can also remove the commented out function if youd like @spencerschrock

**Comment by spencerschrock** [comment]
> updated the test to ignore the date, since the seconds will always be off, and to check for "unknown" commitSHA. should pass unit-tests now! i can also remove the commented out function if youd like @spencerschrock

Which function?

This is what I'm seeing for an empty repo:
```console
go run main.go --repo=github.com/azeemshaikh38/empty-repo-test --format=json  --show-details | jq
{
  "date": "2023-09-14T14:40:01-07:00",
  "repo": {
    "name": "github.com/azeemshaikh38/empty-repo-test",
    "commit": "HEAD"
  },
  "scorecard": {
    "version": "",
    "commit": "unknown"
  },
  "score": -1,
  "checks": null,
  "metadata": null
}
```
And local seems to work still:
```console
go run main.go --local=. --format json --show-details | jq
{
  "date": "2023-09-14T14:44:21-07:00",
  "repo": {
    "name": "file://.",
    "commit": "HEAD"
  },
  "scorecard": {
    "version": "",
    "commit": "unknown"
  },
  "score": 7.9,
 ...
}
```

can review more in a bit

**Comment by leec94** [comment]
sorry i misspoke. it was just the previous way of doing the check i initially commented out, but then just removed. let me know if there's anything i need to change

File: pkg/scorecard.go [hunk 4]
```diff
@@ -83,7 +86,7 @@ func getRepoCommitHash(r clients.RepoClient) (string, error) {
 	if len(commits) > 0 {
 		return commits[0].SHA, nil
 	}
-	return "", nil
+	return "", errEmptyRepository
```
Comment (spencerschrock) [inline pkg/scorecard.go:89]: optional: can we flip this existing code around for "happy path" readability?

```go
if len(commits) == 0 {
	return "", errEmptyRepository
}
return commits[0].SHA, nil
```

File: pkg/scorecard_test.go [hunk 5]
```diff
@@ -168,8 +180,8 @@ func TestRunScorecard(t *testing.T) {
 				t.Errorf("RunScorecard() error = %v, wantErr %v", err, tt.wantErr)
 				return
 			}
-			if !reflect.DeepEqual(got, tt.want) {
-				t.Errorf("RunScorecard() got = %v, want %v", got, tt.want)
+			if !cmp.Equal(got, tt.want, cmpopts.IgnoreFields(ScorecardResult{}, "Date")) { //nolint:govet
+				t.Errorf("expected %v, got %v", got, cmp.Diff(tt.want, got, cmpopts.IgnoreFields(ScorecardResult{}, "Date"))) //nolint:lll
```
Comment (spencerschrock) [inline pkg/scorecard_test.go:184]: one thing you can do to reduce duplication and the line length is saving the `opt`.
```go
ignoreDate := cmpopts.IgnoreFields(ScorecardResult{}, "Date")
if !cmp.Equal(got, tt.want, ignoreDate) {
    t.Errorf("expected %v, got %v", got, cmp.Diff(tt.want, got, ignoreDate))
}
```

ps. not sure what the govet linter was for, doesn't seem to be triggering anymore? can probably be removed

**Review by spencerschrock — APPROVED** [review]

File: pkg/scorecard.go (hunk 4 above)
Comment (leec94) [reply pkg/scorecard.go]: done

File: pkg/scorecard_test.go (hunk 5 above)
Comment (leec94) [reply pkg/scorecard_test.go]: done. thanks for these suggestions and approving the PR!

**Comment by leec94** [comment]
is the code coverage check a blocker to getting this merged?

**Comment by spencerschrock** [comment]
> is the code coverage check a blocker to getting this merged?

No. It's an unfortunate side-effect of our current coverage uploading process. Which we should probably spend some time fixing, as it can be confusing.

**Comment by leec94** [comment]
if you think it's suitable id be happy to help with the code coverage improvements too

**Comment by spencerschrock** [comment]
> if you think it's suitable id be happy to help with the code coverage improvements too

Ah sorry, I had already started working on this yesterday. Just saw this now when I went to link to your comment from #3492

**Comment by leec94** [comment]
no problem! i'd like to keep up helping out tho so if you have suggestions on other issues to look at, that would be great
